File: android/app/src/main/java/com/buildapp/photos/ui/GalleryViewModel.kt
package com.buildapp.photos.ui

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.buildapp.photos.api.IdsBody
import com.buildapp.photos.api.MediaItem
import com.buildapp.photos.api.PhotoApi
import com.buildapp.photos.data.DebugHooks
import com.buildapp.photos.data.SettingsRepository
import com.buildapp.photos.data.Uploader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.intOrNull
import kotlinx.coroutines.withContext

enum class Filter(val label: String) {
    ALL("All"), PHOTOS("Photos"), VIDEOS("Videos"), FAVORITES("Favorites"),
    UPLOADS("Uploads"), ARCHIVED("Archived"), UNKNOWN("Unknown"), TRASH("Trash"),
    LOCKED("Locked folder");

    companion object {
        fun byName(name: String?): Filter? = entries.firstOrNull { it.name.equals(name, ignoreCase = true) }
    }
}

/** Progress of a manual pick-and-upload; shown under the filter row. */
data class UploadProgress(
    val done: Int = 0,
    val total: Int = 0,
    val sent: Int = 0,
    val duplicates: Int = 0,
    val failed: Int = 0,
    val running: Boolean = true,
) {
    val summary: String
        get() = buildString {
            append("Uploaded $sent of $total")
            if (duplicates > 0) append(", $duplicates already in library")
            if (failed > 0) append(", $failed failed")
        }
}

data class GalleryState(
    val items: List<MediaItem> = emptyList(),
    val loading: Boolean = false,
    val error: String? = null,
    val page: Int = 0,
    val total: Int = 0,
    val serverUrl: String = "",
    val itemsIndexed: Int = 0,
    val filter: Filter = Filter.ALL,
    val query: String = "",
    val semantic: Boolean = false,
    val memories: com.buildapp.photos.api.Memories? = null,
    val upload: UploadProgress? = null,
    /** The server answered 401: sign in (accounts exist) or re-enter the password (Locked folder). */
    val needLogin: Boolean = false,
    val needUnlock: Boolean = false,
    val user: com.buildapp.photos.api.User? = null,
)

class GalleryViewModel(app: Application) : AndroidViewModel(app) {
    private val settings = SettingsRepository(app.applicationContext)
    private val _state = MutableStateFlow(GalleryState())
    val state: StateFlow<GalleryState> = _state.asStateFlow()
    private var api: PhotoApi? = null
    private var endReached = false
    // Bumped whenever the list is reset (refresh, filter, query, server). A
    // page that was in flight for an older generation is dropped instead of
    // being spliced into the new list.
    private var gen = 0

    init {
        viewModelScope.launch {
            val url = settings.serverUrl.first()
            val initial = Filter.byName(DebugHooks.initialFilter) ?: Filter.ALL
            _state.value = _state.value.copy(serverUrl = url, filter = initial)
            api = PhotoApi.create(url)
            refresh()
            refreshAuth()
        }
    }

    /**
     * Reload from page 1. The current tiles stay on screen until the fresh
     * first page arrives (no blank flash), then get replaced.
     */
    fun refresh() {
        gen++
        endReached = false
        _state.value = _state.value.copy(page = 0, error = null, loading = false)
        loadNext()
        viewModelScope.launch {
            try {
                val h = api?.health() ?: return@launch
                _state.value = _state.value.copy(itemsIndexed = h.itemsIndexed)
            } catch (_: Exception) {}
        }
        viewModelScope.launch {
            try {
                val mem = api?.memories()
                _state.value = _state.value.copy(memories = mem)
            } catch (_: Exception) {}
        }
    }

    fun loadNext() {
        val api = api ?: return
        val s = _state.value
        if (s.loading || endReached) return
        _state.value = s.copy(loading = true)
        val g = gen
        viewModelScope.launch {
            try {
                val nextPage = s.page + 1
                val kind = when (s.filter) {
                    Filter.PHOTOS -> "photo"
                    Filter.VIDEOS -> "video"
                    else -> null
                }
                val fav = if (s.filter == Filter.FAVORITES) 1 else null
                val trashed = if (s.filter == Filter.TRASH) 1 else null
                val archived = if (s.filter == Filter.ARCHIVED) 1 else null
                val q = s.query.takeIf { it.isNotBlank() }
                // Main feeds show only items with a real capture date; the
                // Unknown chip shows the rest (mirrors the web client).
                val undated = if (s.filter == Filter.UNKNOWN) 1 else null
                val dated = if (q == null && s.filter in listOf(Filter.ALL, Filter.PHOTOS, Filter.VIDEOS)) 1 else null
                // Uploads: the phone-backup album, newest upload first.
                val album = if (s.filter == Filter.UPLOADS) "uploads" else null
                val sort = if (s.filter == Filter.UPLOADS) "added" else null
                val locked = if (s.filter == Filter.LOCKED) 1 else null
                // Semantic search is library-wide and ignores every filter, so
                // it never applies to the Uploads view (a name search does).
                val semantic = s.semantic && q != null && s.filter != Filter.UPLOADS
                val resp = if (semantic) {
                    // Semantic search returns top-k by CLIP similarity (single page).
                    api.searchSemantic(q = q!!, topK = 200)
                } else api.media(
                    page = nextPage, perPage = 80,
                    kind = kind, album = album, q = q,
                    favorites = fav, trashed = trashed, archived = archived,
                    dated = dated, undated = undated, sort = sort, locked = locked,
                )
                if (g != gen) return@launch   // list was reset while this page was in flight
                if (semantic || resp.items.size < resp.perPage) endReached = true
                _state.update { cur ->
                    cur.copy(
                        // Uploads land at the top of the feed while the user
                        // scrolls, shifting offset pages; a repeated id would
                        // be a duplicate LazyGrid key and crash the grid.
                        items = if (nextPage == 1) resp.items else (cur.items + resp.items).distinctBy { it.id },
                        page = nextPage,
                        total = resp.total,
                        loading = false,
                        error = null,
                    )
                }
            } catch (e: Exception) {
                if (g != gen) return@launch
                val code = (e as? retrofit2.HttpException)?.code()
                _state.update {
                    it.copy(
                        loading = false, error = e.message ?: "error",
                        // 401 on the Locked folder means "password again"; anywhere else, sign in.
                        needUnlock = code == 401 && s.filter == Filter.LOCKED,
                        needLogin = code == 401 && s.filter != Filter.LOCKED,
                    )
                }
            }
        }
    }

    fun setFilter(filter: Filter) {
        if (_state.value.filter == filter) return
        gen++
        endReached = false
        _state.value = _state.value.copy(filter = filter, items = emptyList(), page = 0, error = null)
        loadNext()
    }

    fun setQuery(q: String) {
        if (_state.value.query == q) return
        gen++
        endReached = false
        _state.value = _state.value.copy(query = q, items = emptyList(), page = 0, error = null)
        loadNext()
    }

    fun setSemantic(semantic: Boolean) {
        if (_state.value.semantic == semantic) return
        gen++
        endReached = false
        _state.value = _state.value.copy(semantic = semantic, items = emptyList(), page = 0, error = null)
        if (_state.value.query.isNotBlank()) loadNext()
    }

    /**
     * Manual upload of picked items. Runs in the ViewModel scope so a
     * rotation or a dismissed snackbar no longer cancels it mid-transfer;
     * one request per file so a failure costs one file, not the batch.
     */
    fun uploadPicked(uris: List<Uri>) {
        if (uris.isEmpty() || _state.value.upload?.running == true) return
        val url = _state.value.serverUrl
        _state.value = _state.value.copy(upload = UploadProgress(total = uris.size))
        viewModelScope.launch {
            var sent = 0
            var dups = 0
            var failed = 0
            withContext(Dispatchers.IO) {
                Uploader.uploadAll(getApplication(), url, uris) { done, total, r ->
                    if (!r.ok) failed++ else if (r.duplicate) dups++ else sent++
                    // Runs on the IO thread while the main thread also writes
                    // _state; update{} is atomic, a read-copy-write is not.
                    _state.update {
                        it.copy(upload = UploadProgress(done, total, sent, dups, failed, running = done < total))
                    }
                }
            }
            _state.value = _state.value.copy(
                upload = UploadProgress(uris.size, uris.size, sent, dups, failed, running = false),
            )
            // Show the user where the files went.
            if (sent > 0 || dups > 0) {
                if (_state.value.filter == Filter.UPLOADS) refresh() else setFilter(Filter.UPLOADS)
            }
        }
    }

    fun clearUpload() {
        _state.value = _state.value.copy(upload = null)
    }

    fun toggleFavorite(item: MediaItem) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                val r = api.toggleFavorite(item.id)
                _state.value = _state.value.copy(
                    items = _state.value.items.map {
                        if (it.id == item.id) it.copy(favorite = if (r.favorite) 1 else 0) else it
                    }
                )
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    private fun actionFailed(e: Exception) {
        val msg = if (e.message?.contains("403") == true)
            "Rejected: library is read-only" else "Action failed: ${e.message}"
        android.widget.Toast.makeText(getApplication(), msg, android.widget.Toast.LENGTH_SHORT).show()
    }

    fun trash(item: MediaItem) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                api.trash(item.id)
                _state.value = _state.value.copy(
                    items = _state.value.items.filterNot { it.id == item.id },
                    total = _state.value.total - 1,
                )
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    fun archive(item: MediaItem) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                api.archive(item.id)
                _state.value = _state.value.copy(
                    items = _state.value.items.filterNot { it.id == item.id },
                    total = _state.value.total - 1,
                )
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    fun restore(item: MediaItem) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                api.restore(item.id)
                _state.value = _state.value.copy(
                    items = _state.value.items.filterNot { it.id == item.id },
                )
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    fun rotate(item: MediaItem, degrees: Int = 90) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                val resp = api.rotate(item.id, degrees)
                // The server bumps edit_version and returns it; take its value
                // so the versioned (immutably cached) thumb URL changes and the
                // tile reloads, and stays in step with what a refresh returns.
                val fromServer = (resp["edit_version"] as? JsonPrimitive)?.intOrNull
                _state.value = _state.value.copy(
                    items = _state.value.items.map {
                        if (it.id == item.id) it.copy(editVersion = fromServer ?: (it.editVersion + 1)) else it
                    }
                )
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    /** Who is signed in (null while the server is open or after sign-out). */
    fun refreshAuth() {
        viewModelScope.launch {
            try {
                val a = api?.authState() ?: return@launch
                _state.update { it.copy(user = a.user) }
            } catch (_: Exception) {}
        }
    }

    fun login(name: String, password: String, onResult: (String?) -> Unit) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                val r = api.login(com.buildapp.photos.api.LoginBody(name, password))
                settings.setSession(r.token, r.user.name)
                _state.update { it.copy(needLogin = false, user = r.user, error = null) }
                onResult(null)
                refresh()
            } catch (e: Exception) {
                onResult(if ((e as? retrofit2.HttpException)?.code() == 401) "Wrong name or password" else (e.message ?: "error"))
            }
        }
    }

    fun logout() {
        viewModelScope.launch {
            try { api?.logout() } catch (_: Exception) {}
            settings.setSession(null, null)
            _state.update { it.copy(user = null) }
            refresh()
        }
    }

    /** Open the Locked folder for this session with the account password. */
    fun unlockLocked(password: String, onResult: (Boolean) -> Unit) {
        val api = api ?: return
        viewModelScope.launch {
            val ok = try { api.unlockLocked(com.buildapp.photos.api.PasswordBody(password)); true } catch (_: Exception) { false }
            if (ok) _state.update { it.copy(needUnlock = false) }
            onResult(ok)
            if (ok && _state.value.filter == Filter.LOCKED) refresh()
        }
    }

    fun clearAuthPrompts() { _state.update { it.copy(needLogin = false, needUnlock = false) } }

    /** Move an item into (or out of) my Locked folder; it leaves the current list either way. */
    fun setLocked(item: MediaItem, locked: Boolean) {
        val api = api ?: return
        viewModelScope.launch {
            try {
                if (locked) api.lockMedia(IdsBody(listOf(item.id))) else api.unlockMedia(IdsBody(listOf(item.id)))
                _state.update { s -> s.copy(items = s.items.filterNot { it.id == item.id }) }
            } catch (e: Exception) { actionFailed(e) }
        }
    }

    fun setServerUrl(url: String) {
        viewModelScope.launch {
            gen++   // a page still in flight from the old server must not land in the new list
            settings.setServerUrl(url)
            api = PhotoApi.create(url)
            _state.value = GalleryState(serverUrl = url)
            refresh()
        }
    }

    fun rescan() {
        viewModelScope.launch {
            try { api?.rescan() } catch (_: Exception) {}
            refresh()
        }
    }
}

File: ios/Chitra/UI/GalleryViewModel.swift
import Foundation
import SwiftUI
import Photos
import PhotosUI

@MainActor
final class GalleryViewModel: ObservableObject {
    @Published private(set) var items: [MediaItem] = []
    @Published private(set) var loading = false
    @Published private(set) var error: String?
    @Published private(set) var total = 0
    @Published private(set) var itemsIndexed = 0
    @Published private(set) var memories: Memories?
    @Published private(set) var serverURL: String
    @Published var filter: Filter = .all
    @Published var query = ""
    @Published var semantic = false
    @Published var upload: UploadProgress?
    /// One-off message, surfaced as a brief overlay.
    @Published var notice: String?
    /// Selection mode, driven by the library's "Select" button.
    @Published var isSelecting = false
    @Published var selection: Set<String> = []

    private let settings = SettingsStore.shared
    private var api: PhotoAPI
    private var page = 0
    private var endReached = false
    /// Bumped whenever the list is reset (refresh, filter, query, server). A
    /// page that was in flight for an older generation is dropped instead of
    /// being spliced into the new list.
    private var generation = 0
    private var searchDebounce: Task<Void, Never>?

    init() {
        let url = settings.serverURL
        serverURL = url
        api = PhotoAPI(baseUrl: url)
        filter = DebugHooks.initialFilter ?? .all
        refresh()
    }

    // MARK: - Loading

    /// Reload from page 1. The current tiles stay on screen until the fresh
    /// first page arrives (no blank flash), then get replaced.
    func refresh() {
        generation += 1
        endReached = false
        page = 0
        error = nil
        loading = false
        loadNext()

        Task { [api] in
            if let health = try? await api.health() { itemsIndexed = health.itemsIndexed }
        }
        Task { [api] in
            memories = try? await api.memories()
        }
    }

    func loadNext() {
        guard !loading, !endReached else { return }
        loading = true
        let generationAtStart = generation
        let nextPage = page + 1
        let filter = filter
        let query = query.trimmingCharacters(in: .whitespaces)
        let semantic = semantic

        Task { [api] in
            do {
                let kind: String? = {
                    switch filter {
                    case .photos: return "photo"
                    case .videos: return "video"
                    default: return nil
                    }
                }()
                let q = query.isEmpty ? nil : query
                // Main feeds show only items with a real capture date; the
                // Unknown chip shows the rest (mirrors the web client).
                let undated = filter == .unknown ? 1 : nil
                let dated = (q == nil && [.all, .photos, .videos].contains(filter)) ? 1 : nil
                // Uploads: the phone-backup album, newest upload first.
                let album = filter == .uploads ? "uploads" : nil
                let sort = filter == .uploads ? "added" : nil
                // Semantic search is library-wide and ignores every filter, so
                // it never applies to the Uploads view (a name search does).
                let useSemantic = semantic && q != nil && filter != .uploads

                let response: MediaPage = useSemantic
                    ? try await api.searchSemantic(q: q!, topK: 200)
                    : try await api.media(
                        page: nextPage, perPage: 80, kind: kind, album: album, q: q,
                        favorites: filter == .favorites ? 1 : nil,
                        trashed: filter == .trash ? 1 : nil,
                        archived: filter == .archived ? 1 : nil,
                        dated: dated, undated: undated, sort: sort)

                guard generationAtStart == generation else { return }  // list was reset mid-flight
                if useSemantic || response.items.count < response.perPage { endReached = true }
                // Uploads land at the top of the feed while the user scrolls,
                // shifting offset pages; a repeated id would be a duplicate
                // SwiftUI identity and drop rows out of the grid.
                items = nextPage == 1 ? response.items : dedupe(items + response.items)
                page = nextPage
                total = response.total
                loading = false
                error = nil
            } catch {
                guard generationAtStart == generation else { return }
                loading = false
                self.error = error.localizedDescription
            }
        }
    }

    private func dedupe(_ list: [MediaItem]) -> [MediaItem] {
        var seen = Set<String>()
        return list.filter { seen.insert($0.id).inserted }
    }

    /// True once the grid has scrolled within 20 tiles of the end.
    func loadMoreIfNeeded(currentItem: MediaItem) {
        guard let index = items.firstIndex(of: currentItem) else { return }
        if index >= items.count - 20 { loadNext() }
    }

    private func resetAndLoad() {
        endSelection()
        generation += 1
        endReached = false
        items = []
        page = 0
        error = nil
        loading = false
        loadNext()
    }

    func setFilter(_ newFilter: Filter) {
        guard filter != newFilter else { return }
        filter = newFilter
        resetAndLoad()
    }

    /// Typing re-runs the search after a short pause, rather than firing one
    /// request per keystroke the way a plain text-change listener would.
    func setQuery(_ newQuery: String) {
        guard query != newQuery else { return }
        query = newQuery
        searchDebounce?.cancel()
        searchDebounce = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            self?.resetAndLoad()
        }
    }

    func setSemantic(_ on: Bool) {
        guard semantic != on else { return }
        semantic = on
        if !query.trimmingCharacters(in: .whitespaces).isEmpty { resetAndLoad() }
    }

    func setServerURL(_ url: String) {
        generation += 1   // a page still in flight from the old server must not land in the new list
        settings.setServerURL(url)
        serverURL = settings.serverURL
        api = PhotoAPI(baseUrl: serverURL)
        items = []
        page = 0
        total = 0
        endReached = false
        error = nil
        memories = nil
        refresh()
    }

    func rescan() {
        Task { [api] in
            try? await api.rescan()
            refresh()
        }
    }

    // MARK: - Item actions

    private func actionFailed(_ error: Error) {
        if let apiError = error as? APIError, apiError.isForbidden {
            notice = "Rejected: library is read-only"
        } else {
            notice = "Action failed: \(error.localizedDescription)"
        }
    }

    func toggleFavorite(_ item: MediaItem) {
        Task { [api] in
            do {
                let response = try await api.toggleFavorite(item.id)
                update(item.id) { $0.favorite = response.favorite ? 1 : 0 }
            } catch { actionFailed(error) }
        }
    }

    func trash(_ item: MediaItem) {
        Task { [api] in
            do {
                try await api.trash(item.id)
                remove(item.id)
            } catch { actionFailed(error) }
        }
    }

    func archive(_ item: MediaItem) {
        Task { [api] in
            do {
                try await api.archive(item.id)
                remove(item.id)
            } catch { actionFailed(error) }
        }
    }

    func restore(_ item: MediaItem) {
        Task { [api] in
            do {
                try await api.restore(item.id)
                items.removeAll { $0.id == item.id }
            } catch { actionFailed(error) }
        }
    }

    func rotate(_ item: MediaItem, degrees: Int = 90) {
        Task { [api] in
            do {
                // The server bumps edit_version and returns it; take its value
                // so the versioned (immutably cached) thumb URL changes and the
                // tile reloads, and stays in step with what a refresh returns.
                let fromServer = try await api.rotate(item.id, degrees: degrees)
                update(item.id) { $0.editVersion = fromServer ?? ($0.editVersion + 1) }
            } catch { actionFailed(error) }
        }
    }

    // MARK: - Selection

    func beginSelection() {
        isSelecting = true
        selection = []
    }

    func endSelection() {
        isSelecting = false
        selection = []
    }

    func toggleSelection(_ item: MediaItem) {
        if selection.contains(item.id) { selection.remove(item.id) } else { selection.insert(item.id) }
    }

    var selectedItems: [MediaItem] { items.filter { selection.contains($0.id) } }

    /// Trash, restore and permanent delete have batch endpoints. Favourite and
    /// archive do not, so those walk the selection one call at a time.
    func trashSelected() {
        runBatch { api, ids in try await api.batchTrash(ids) } thenRemove: { true }
    }

    func restoreSelected() {
        runBatch { api, ids in try await api.batchRestore(ids) } thenRemove: { true }
    }

    func deleteSelected() {
        runBatch { api, ids in try await api.batchDelete(ids) } thenRemove: { true }
    }

    func favoriteSelected() {
        let chosen = selectedItems
        endSelection()
        Task { [api] in
            for item in chosen where !item.isFavorite {
                if let response = try? await api.toggleFavorite(item.id) {
                    update(item.id) { $0.favorite = response.favorite ? 1 : 0 }
                }
            }
        }
    }

    func archiveSelected() {
        let ids = Array(selection)
        endSelection()
        Task { [api] in
            for id in ids {
                do {
                    try await api.archive(id)
                    remove(id)
                } catch { actionFailed(error) }
            }
        }
    }

    private func runBatch(_ call: @escaping (PhotoAPI, [String]) async throws -> Void,
                          thenRemove: @escaping () -> Bool) {
        let ids = Array(selection)
        guard !ids.isEmpty else { return }
        endSelection()
        Task { [api] in
            do {
                try await call(api, ids)
                if thenRemove() {
                    items.removeAll { ids.contains($0.id) }
                    total = max(0, total - ids.count)
                }
            } catch { actionFailed(error) }
        }
    }

    private func update(_ id: String, _ change: (inout MediaItem) -> Void) {
        guard let index = items.firstIndex(where: { $0.id == id }) else { return }
        change(&items[index])
    }

    private func remove(_ id: String) {
        items.removeAll { $0.id == id }
        total = max(0, total - 1)
    }

    // MARK: - Manual upload

    /// Manual upload of picked items. Runs detached from any view's lifetime
    /// so leaving the screen no longer cancels it mid-transfer; one request
    /// per file so a failure costs one file, not the batch.
    func uploadPicked(_ picked: [PhotosPickerItem]) {
        guard !picked.isEmpty, upload?.running != true else { return }
        let serverUrl = serverURL
        upload = UploadProgress(total: picked.count)
        Task {
            let device = DeviceInfo.name()
            var sent = 0, duplicates = 0, failed = 0
            for (index, item) in picked.enumerated() {
                let result = await Self.uploadPickedItem(item, serverUrl: serverUrl, device: device)
                if !result.ok { failed += 1 } else if result.duplicate { duplicates += 1 } else { sent += 1 }
                upload = UploadProgress(done: index + 1, total: picked.count, sent: sent,
                                        duplicates: duplicates, failed: failed,
                                        running: index + 1 < picked.count)
            }
            let finished = UploadProgress(done: picked.count, total: picked.count, sent: sent,
                                          duplicates: duplicates, failed: failed, running: false)
            upload = finished
            notice = finished.summary
            // Show the user where the files went.
            if sent > 0 || duplicates > 0 {
                if filter == .uploads { refresh() } else { setFilter(.uploads) }
            }
        }
    }

    /// The picker hands back an opaque item. When the library is authorized it
    /// carries a PHAsset identifier and we can stream the original; otherwise
    /// the bytes come through the transfer API into a temp file.
    private static func uploadPickedItem(_ item: PhotosPickerItem, serverUrl: String, device: String) async -> Uploader.FileResult {
        if let identifier = item.itemIdentifier, let asset = DeviceMedia.asset(withId: identifier) {
            let info = DeviceMedia.resourceInfo(for: asset)
            return await Uploader.upload(asset: asset, name: info.name, serverUrl: serverUrl,
                                         source: Uploader.Source(folder: nil, device: device))
        }
        guard let data = try? await item.loadTransferable(type: Data.self) else {
            return Uploader.FileResult(name: "picked", ok: false, error: "could not read item")
        }
        let ext = item.supportedContentTypes.first?.preferredFilenameExtension ?? "jpg"
        let name = "IMG-\(Int(Date().timeIntervalSince1970 * 1000)).\(ext)"
        let temporary = FileManager.default.temporaryDirectory.appendingPathComponent(name)
        defer { try? FileManager.default.removeItem(at: temporary) }
        guard (try? data.write(to: temporary)) != nil else {
            return Uploader.FileResult(name: name, ok: false, error: "could not stage item")
        }
        return await Uploader.upload(fileAt: temporary, name: name, serverUrl: serverUrl,
                                     source: Uploader.Source(folder: nil, device: device))
    }

    func clearUpload() { upload = nil }
}
